src/main/java/io/fabric8/maven/docker/config/BuildImageConfiguration.java

```diff
@@ -548,20 +576,48 @@ private void initDockerFileFile(Logger log) {
 
     private File findDockerFileFile(Logger log) {
         if (dockerFile != null) {
+            if (EnvUtil.isWindows() && !EnvUtil.isValidWindowsFileName(dockerFile)) {
+                throw new IllegalArgumentException(String.format("Invalid Windows file name %s for <dockerFile>", dockerFile));
+            }
+
             File dFile = new File(dockerFile);
-            if (dockerFileDir == null) {
+            if (dockerFileDir == null && contextDir == null) {
                 return dFile;
             } else {
-                if (dFile.isAbsolute()) {
-                    throw new IllegalArgumentException("<dockerFile> can not be absolute path if <dockerFileDir> also set.");
+                if(contextDir != null) {
+                    if(dockerFileDir != null) {
+                        log.warn("using contextDir as it has more precedence for Dockerfile");
+                    }
+
+                    if (dFile.isAbsolute()) {
+                        try {
+                            Files.copy(dFile.toPath(), Paths.get(contextDir));
+                        } catch (IOException exception) {
+                            log.warn("Error during copying Dockerfile to contextDir: ", exception.getMessage());
+                        }
+                        return new File(contextDir, "Dockerfile");
+                    }
+                    return new File(contextDir, dockerFile);
                 }
-                if (EnvUtil.isWindows() && !EnvUtil.isValidWindowsFileName(dockerFile)) {
-                    throw new IllegalArgumentException(String.format("Invalid Windows file name %s for <dockerFile>", dockerFile));
+
+                if (dockerFileDir != null) {
+                    if (dFile.isAbsolute()) {
+                        throw new IllegalArgumentException("<dockerFile> can not be absolute path if <dockerFileDir> also set.");
+                    }
+                    log.warn("dockerFileDir parameter is deprecated, please migrate to contextDir");
+                    return new File(dockerFileDir, dockerFile);
                 }
-                return new File(dockerFileDir, dockerFile);
             }
         }
 
+
+        if (contextDir != null) {
```

Comment:
I recommend to not duplicate that check but just add it outside the conditionals (and use the comment as mentioned above)